src/dtls.c

```diff
@@ -290,60 +374,530 @@ static int TlsSessionIdIsValid(WOLFSSL* ssl, WolfSSL_ConstVector sessionID,
-    ret = TlsSessionIdIsValid(ssl, ch->sessionId, isValid);
+    ret = TlsSessionIdIsValid(ssl, ch->sessionId, resume);
+    return ret;
+}
+#endif /* WOLFSSL_DTLS13 || WOLFSSL_DTLS_NO_HVR_ON_RESUME */
+
+#ifdef WOLFSSL_DTLS13
+static int TlsCheckSupportedVersion(const WOLFSSL* ssl,
+        WolfSSL_CH* ch, byte *isTls13)
+{
+    WolfSSL_ConstVector tlsxSupportedVersions;
+    int ret;
+    ProtocolVersion pv = ssl->version;
+    int tlsxFound;
+
+    ret = TlsxFindByType(&tlsxSupportedVersions, TLSX_SUPPORTED_VERSIONS,
+                         ch->extension, &tlsxFound);
+    if (ret != 0)
+        return ret;
+    if (!tlsxFound) {
+        *isTls13 = 0;
+        return 0;
+    }
+    ret = TLSX_SupportedVersions_Parse(ssl, tlsxSupportedVersions.elements,
+            (word16)tlsxSupportedVersions.size, client_hello, &pv, NULL, NULL);
+    if (ret != 0)
+        return ret;
+    if (IsAtLeastTLSv1_3(pv))
+        *isTls13 = 1;
+    else
+        *isTls13 = 0;
+
+    return 0;
+}
+
+static int CopySupportedGroup(TLSX* src, TLSX** dst, void* heap)
+{
+    TLSX* extension;
+    int ret;
+
+    extension = TLSX_Find(src, TLSX_SUPPORTED_GROUPS);
+    if (extension != NULL) {
+        SupportedCurve* curve;
+        curve = (SupportedCurve*)extension->data;
+        for (curve = (SupportedCurve*)extension->data; curve != NULL;
+                curve = curve->next) {
+            ret = TLSX_UseSupportedCurve(dst, curve->name, heap);
+            if (ret != WOLFSSL_SUCCESS)
+                return MEMORY_E;
+        }
+    }
+
+    return 0;
+}
+#endif
+
+#if defined(WOLFSSL_DTLS13) && \
+        (!defined(NO_PSK) || defined(HAVE_SESSION_TICKET))
+/* Very simplified version of CheckPreSharedKeys to find the current suite */
+static void FindPskSuiteFromExt(const WOLFSSL* ssl, TLSX* extensions,
+                                PskInfo* pskInfo, Suites* suites)
+{
+    TLSX* pskExt = TLSX_Find(extensions, TLSX_PRE_SHARED_KEY);
+    PreSharedKey* current;
+    int i;
+    int ret;
+
+    if (pskExt == NULL)
+        return;
+
+    for (i = 0; i < suites->suiteSz; i += 2) {
+        for (current = (PreSharedKey*)pskExt->data; current != NULL;
+                current = current->next) {
+#ifdef HAVE_SESSION_TICKET
+            {
+                /* Decode the identity. */
+                switch (current->decryptRet) {
+                    case PSK_DECRYPT_NONE:
+                        ret = DoClientTicket_ex(ssl, current);
+                        break;
+                    case PSK_DECRYPT_OK:
+                        ret = WOLFSSL_TICKET_RET_OK;
+                        break;
+                    case PSK_DECRYPT_CREATE:
+                        ret = WOLFSSL_TICKET_RET_CREATE;
+                        break;
+                    case PSK_DECRYPT_FAIL:
+                    default:
+                        ret = WOLFSSL_TICKET_RET_REJECT;
+                        break;
+                }
+                if (ret == WOLFSSL_TICKET_RET_OK) {
```

Comment:
should we check for  `WOLFSSL_TICKET_RET_CREATE`  as well?

Reply:
```
 * @return  WOLFSSL_TICKET_RET_CREATE when successful and a new ticket is to
 *          be created for TLS 1.2 and below.
```
This is our documentation regarding `WOLFSSL_TICKET_RET_CREATE`. It is also not used in the stateful `DoPreSharedKeys`.